lib/THC/THCApply.cuh

```diff
@@ -0,0 +1,694 @@
+#ifndef THC_APPLY_INC
+#define THC_APPLY_INC
+
+#include "THCTensorCopy.h"
+#include "THCReduceApplyUtils.cuh"
+
+//
+// This file contains pointwise operation functions and kernels that
+// work on both contiguous and non-contiguous tensor arguments of
+// arbitrary (up to MAX_CUTORCH_DIMS) dimensioned arguments without
+// copying or temporary storage.
+//
+
+// Maximum elements per thread that we will handle in our apply kernel
+#define THC_APPLY_ELEM_PER_THREAD 8
+
+// Threads per block for our apply kernel
+#define THC_APPLY_THREADS_PER_BLOCK 32 * 16
+
+// Called when we are copying into an overlapping index `dst`, but
+// we don't care which writer wins. Hacky but it works.
+void THCudaTensor_copyIgnoringOverlaps(THCState* state,
+                                       THCudaTensor* dst,
+                                       THCudaTensor* src);
+
+template <typename IndexType>
+__device__ __forceinline__ IndexType
+getStartLinearIndex(const IndexType blockId) {
+  return blockId * THC_APPLY_THREADS_PER_BLOCK * THC_APPLY_ELEM_PER_THREAD;
+}
+
+template <typename IndexType>
+__device__ __forceinline__ IndexType
+getEndLinearIndex(const IndexType blockId, const IndexType totalElements) {
+  IndexType endLinearId =
+    (blockId + 1) * THC_APPLY_THREADS_PER_BLOCK * THC_APPLY_ELEM_PER_THREAD;
+  endLinearId = endLinearId < totalElements ? endLinearId : totalElements;
+
+  return endLinearId;
+}
+
+template <typename Op, typename IndexType, int ADims>
+#if __CUDA_ARCH__ >= 350
+__launch_bounds__(32 * 16, 4)
+#endif
+__global__ void
+THCudaTensor_pointwiseApply1(TensorInfo<IndexType> a,
+                             IndexType totalElements,
+                             Op op) {
+  const IndexType blockId = getLinearBlockId<IndexType>();
+  const IndexType startLinearIndex =
+    getStartLinearIndex<IndexType>(blockId);
+
+  if (blockId == getLastLinearBlockId<IndexType>()) {
```

Comment:
for massive tensors with >536 million elements, virtual (with stride 0) or physical, this logic is wrong, fixing...